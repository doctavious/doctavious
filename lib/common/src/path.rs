use std::ffi::{OsStr, OsString};
use std::path::PathBuf;

pub fn append_to_path(p: impl Into<OsString>, s: impl AsRef<OsStr>) -> PathBuf {
    let mut p = p.into();
    p.push(s);
    p.into()
}
