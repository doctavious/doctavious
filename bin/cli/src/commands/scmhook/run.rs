use std::path::PathBuf;

use clap::Parser;
use doctavious_cli::cmd::scm_hooks::run::run;
use doctavious_cli::CliResult;

/// Execute commands/scripts associated to the specified hook.
///
/// This is called for every hook managed by doctavious.
/// You can also provide your own hooks that can only be called manually.
#[derive(Parser, Debug)]
#[command()]
pub(crate) struct RunScmHookCommand {
    /// Name of the hook to run
    #[arg(index = 1)]
    pub hook: String,

    /// Path to execute run
    #[arg(long, short)]
    pub cwd: Option<PathBuf>,

    // TODO: can use group = "files" to only allow one to be used?
    /// Run on specified file (repeat for multiple files). takes precedence over --all-files
    #[arg(long, short)]
    pub file: Option<Vec<PathBuf>>,

    /// Run hooks on all files
    #[arg(long, short, action)]
    pub all_files: bool,

    /// Run only specified executions (commands / scripts)
    #[arg(long = "executions", short = 'e')]
    pub run_only_executions: Option<Vec<String>>,

    /// Force execution of commands that can be skipped
    #[arg(long, short, action)]
    pub force: bool,
}

pub(crate) fn execute(command: RunScmHookCommand) -> CliResult<Option<String>> {
    let path = command.cwd.unwrap_or(std::env::current_dir()?);

    // TODO: turn all_files / files into an enum

    run(
        &path,
        &command.hook,
        command.all_files,
        command.file.unwrap_or_default(),
        command.run_only_executions.unwrap_or_default(),
        command.force,
    )?;

    Ok(None)
}
