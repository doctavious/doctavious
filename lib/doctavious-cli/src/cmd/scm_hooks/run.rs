use std::path::{Path, PathBuf};

use scm::drivers::{Scm, ScmRepository};
use scm::errors::ScmError;
use thiserror::Error;
use tracing::{info, warn};

use crate::cmd::scm_hooks::ensure_hooks;
use crate::cmd::scm_hooks::runner::{
    ScmHookRunner, ScmHookRunnerOptions, ScmHookRunnerOutcome, ScmHookRunnerResult,
};
use crate::errors::{CliResult, DoctaviousCliError};
use crate::settings::{load_settings, SettingErrors, Settings};

#[remain::sorted]
#[derive(Debug, Error)]
pub enum DoctaviousScmHookRunError {
    #[error("{0}")]
    ConflictingOptions(String),
}

#[remain::sorted]
pub enum ScmHookRunFiles {
    All,
    Specific(Vec<PathBuf>),
}

pub fn run(
    cwd: &Path,
    hook_name: &str,
    files: Option<ScmHookRunFiles>,
    run_only_executions: Vec<String>,
    synchronize_hooks: bool,
    force: bool,
) -> CliResult<()> {
    let settings: Settings = load_settings(cwd)?;
    let Some(scm_settings) = settings.scmhook_settings else {
        return Err(DoctaviousCliError::SettingError(
            SettingErrors::SectionNotFound(
                "Either edit `scm` settings in doctavious configuration or use `scmhook add`"
                    .to_string(),
            ),
        ));
    };

    let scm = Scm::get(cwd)?;
    if synchronize_hooks {
        // TODO: do we want to return error?
        if ensure_hooks(&scm_settings, &scm, true, force).is_err() {
            warn!("There was a problem synchronizing hooks. Try running 'doctavious scmhook install' manually")
        }
    }

    let Some(hook) = scm_settings.hooks.get(hook_name) else {
        return Err(DoctaviousCliError::ScmError(ScmError::UnsupportedHook(
            hook_name.to_string(),
        )));
    };

    if hook.parallel && hook.stop_on_failure {
        // TODO: return error
        // conflicting options 'piped' and 'parallel' are set to 'true', remove one of this option from hook group
    }

    let files = match files {
        None => vec![],
        Some(f) => match f {
            ScmHookRunFiles::All => scm.all_files()?,
            ScmHookRunFiles::Specific(files) => files,
        },
    };

    let runner = ScmHookRunner::new(ScmHookRunnerOptions {
        cwd,
        hook,
        hook_name: hook_name.to_string(),
        files,
        run_only_executions,
        force,
    });

    let results = runner.run_all();
    print_summary(results);

    Ok(())
}

fn print_summary(results: Vec<ScmHookRunnerResult<ScmHookRunnerOutcome>>) {
    // TODO: have log settings
    // TODO: print summary
    for result in results {
        match result {
            Ok(r) => {
                // TODO: color green
                info!("✔️  {}", "name")
            }
            Err(e) => {
                // TODO: color red
                info!("{}", e)
            }
        }
    }
}
