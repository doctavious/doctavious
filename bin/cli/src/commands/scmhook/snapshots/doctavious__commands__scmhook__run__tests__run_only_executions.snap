---
source: bin/cli/src/commands/scmhook/run.rs
expression: "fs::read_to_string(&temp_path.join(\"backend/src/lib.rs\")).unwrap()"
---
fn main() {
    println!("Hello, world!");
}

